Adding proper accommodation for textScaleFactor in chips, and StadiumBorder border.

gspencergoog opened the pull request:
In order to allow chips to be properly drawn when they expand in size (without
using IntrinsicHeight), I needed a BoxDecoration shape that would be dependent
upon the rendered height of the widget.  This seemed to be pretty generally
useful, so I added a new ShapeDecoration called StadiumBorder.  It uses the minimum dimension to
adjust the BorderRadius of a rounded rect in the shape decoration.

I also converted some uses of BoxShape to be case statements, updated the
chips to use the StadiumBorder decoration, and updated some of the metrics to match
the Material spec.

gspencergoog committed:
Adding BoxShape.pill and proper accomodation for textScaleFactor in chips

In order to allow chips to be properly drawn when they expand in size (without
using IntrinsicHeight), I needed a BoxDecoration shape that would be dependent
upon the rendered height of the widget.  This seemed to be pretty generally
useful, so I added the new BoxShape.pill.  It uses the minimum dimension to
adjust the BorderRadius of a rounded rect in the box decoration.

I also converted some uses of BoxShape to be case statements, updated the
chips to use the pill decoration, and updated some of the metrics to match
the Material spec.

Hixie commented:
How is this not BoxShape.rectangle with a border radius?

Hixie commented:
Rather than adding new BoxShapes to BoxBorder, we should add a new ShapeBorder. Otherwise we're not going to get proper interpolation behaviour. See my work in progress here:
https://github.com/flutter/flutter/compare/master...Hixie:rounded_rectangle_border
https://github.com/flutter/flutter/compare/master...Hixie:circle_border
https://github.com/flutter/flutter/compare/master...Hixie:shape_decoration

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -29,7 +28,8 @@ const TextStyle _kLabelStyle = const TextStyle(
 /// button for deleting the chip.
 ///
-/// Requires one of its ancestors to be a [Material] widget.
+/// Requires one of its ancestors to be a [Material] widget.  [label] may not be
```
Try to avoid starting sentences with a lowercase letter.
Also, single space after periods.

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -57,11 +57,13 @@ class Chip extends StatelessWidget {
     this.deleteIconColor,
-  }) : super(key: key);
+  }) : assert(label != null),
+       labelStyle = labelStyle ?? _kLabelStyle,
```
prefer default argument values

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -135,12 +138,13 @@ class Chip extends StatelessWidget {
       children.add(new GestureDetector(
         onTap: Feedback.wrapForTap(onDeleted, context),
         child: new Tooltip(
+          // TODO(gspencer): Internationalize this text.
```
cc @HansMuller 
link to https://github.com/flutter/flutter/issues/12378

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -135,12 +138,13 @@ class Chip extends StatelessWidget {
           message: deleteButtonTooltipMessage ?? 'Delete "$label"',
           child: new Container(
-            padding: const EdgeInsets.symmetric(horizontal: 4.0),
+            padding: const EdgeInsetsDirectional.only(start: 4.0, end: 4.0),
```
?

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -150,16 +154,22 @@ class Chip extends StatelessWidget {
-          children: children,
-          mainAxisSize: MainAxisSize.min,
+      child: new ConstrainedBox(
+        constraints: const BoxConstraints(minHeight: _kChipHeight),
```
you can just put the constraints on the Container, since you have one already.

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -150,16 +154,22 @@ class Chip extends StatelessWidget {
+          decoration: new BoxDecoration(
+            color: backgroundColor ?? Colors.grey.shade300,
+            shape: BoxShape.pill,
+          ),
```
we should use the new decoration infrastructure, rather than adding a new enum value

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -123,6 +131,21 @@ abstract class BoxBorder extends ShapeBorder {
+  /// supplied rectangle.
+  ///
+  /// The [shape] parameter can only be [BoxShape.pill].
+  static BorderRadius pillRadiusForRect(BoxShape shape, Rect rect) {
```
shouldn't this just be a constructor on BorderRadius?

Hixie commented:
I applaud the comprehensiveness of your patch, at least. :-)

gspencergoog commented:
It's not simply a BoxShape.rectangle with a border radius because the radius depends upon the dimensions of the box, rather than being fixed.  Will your new decorations handle that case?

gspencergoog commented:
I'm going to hold off on doing anything more for this PR until I have a chance to discuss with Ian about the new infrastructure and how that might do what I want here.

Hixie commented:
> It's not simply a BoxShape.rectangle with a border radius because the radius depends upon the dimensions of the box, rather than being fixed. Will your new decorations handle that case?

Sure, it'll handle anything you can do with a Path.

Hixie commented:
BTW, I believe the term of art for this shape is "stadium" rather than "pill" (see the RRect class for precedent).

gspencergoog committed:
Revert pill changes.

gspencergoog committed:
Creating Stadium Border

gspencergoog commented:
PTAL: I incorporated the ShapeBorder and reverted (most of) the pill shape changes (I kept the conversions to switches instead of if's).  I made StadiumBorder, the new ShapeBorder, a separate class because it seemed generally useful.

gspencergoog commented:
Oh, wait, I didn't add tests for the StadiumBorder.  Working on that still.

gspencergoog committed:
Adding stadium border test

gspencergoog committed:
Updating DecorationTween docs

gspencergoog committed:
Adding side named argument to CircleBorder

gspencergoog commented:
OK, PTAL.  There's now a StadiumBorder type that can animate between other border types, and tests for the same.  I've also update the docs for DecorationTween, and added a "side" named parameter to CircleBorder to replace the optional positional parameter.

Hixie commented on examples/flutter_gallery/lib/demo/material/chip_demo.dart:
```diff
@@ -28,7 +28,8 @@ class _ChipDemoState extends State<ChipDemo> {
         avatar: const CircleAvatar(child: const Text('B')),
-        label: const Text('Blueberry')
+        label: const Text('Blueberry'),
+        border: const BorderSide(color: Colors.redAccent, width: 4.0),
```
did you mean to leave this border side? I don't think I've seen chips with a border like this in the material spec.

gspencergoog replied:
Yeah, I wanted to leave it as an example, but you're right that it's not in the spec.  Removed.

gspencergoog committed:
Review changes.

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -135,12 +146,13 @@ class Chip extends StatelessWidget {
       children.add(new GestureDetector(
         onTap: Feedback.wrapForTap(onDeleted, context),
         child: new Tooltip(
+          // TODO(#12378): Internationalize this text.
```
preferred syntax is `TODO(username): Message, https://...`

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -475,16 +475,20 @@ class Border extends BoxBorder {
+              }
+              BoxBorder._paintUniformBorderWithRectangle(canvas, rect, top);
+              break;
+            break;
```
you have two breaks here

Hixie replied:
i'm surprised the analyzer didn't catch this

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1372,12 +1372,15 @@ class RenderPhysicalModel extends _RenderCustomClip<RRect> {
+        final Rect rect = Offset.zero & size;
+        return new RRect.fromRectXY(rect, rect.width / 2, rect.height / 2);
     }
+    assert(false, 'Box shape $_shape not handled.');
```
no need for this assert, the analyzer will catch it. just assert non-null before the switch.

Hixie commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -36,12 +36,22 @@ class BoxConstraintsTween extends Tween<BoxConstraints> {
+/// interpolation between decorations.
 ///
-/// See [Tween] for a discussion on how to use interpolation objects.
+/// For other types of decorations (e.g. [BoxDecoration]), this
```
BoxDecoration is actually a poor example here because ShapeDecoration knows how to lerp to/from BoxDecoration.

gspencergoog replied:
OK... But the only other extenders of Decoration are BoxDecoration and FlutterLogoDecoration.  The rest are all ShapeDecorations already.  I'm just going to remove this paragraph.

Hixie commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -36,12 +36,22 @@ class BoxConstraintsTween extends Tween<BoxConstraints> {
+///
+/// See also:
+///   * [Tween] for a discussion on how to use interpolation objects.
+///   * [ShapeDecoration], [RoundedRectangleBorder], [CircleBorder],
```
...and

gspencergoog replied:
Done.

Hixie commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -254,6 +254,155 @@ void main() {
+      ),
+    );
+
+    // TODO(#12357): Update this test when the font metric bug is fixed to remove the anyOfs.
```
as in the TODO above, please provide a link rather than just a bug number

gspencergoog replied:
Done.  And below.  Sorry, this is best practice internally, but I guess Github and IDEs don't link those automatically.

Hixie commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -254,6 +254,155 @@ void main() {
+      ),
+    );
+
+    // TODO(#12357): Update this test when the font metric bug is fixed to remove the anyOfs.
```
ditto

Hixie commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -254,6 +254,155 @@ void main() {
+      ),
+    );
+
+    // TODO(#12357): Update this test when the font metric bug is fixed to remove the anyOfs.
```
ditto

Hixie commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -254,6 +254,155 @@ void main() {
+      ),
+    );
+
+    // TODO(#12357): Update this test when the font metric bug is fixed to remove the anyOfs.
```
ditto

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -90,6 +95,11 @@ class Chip extends StatelessWidget {
+  /// The border to draw around the chip.
+  ///
+  /// Defaults to [BorderSide.none].
+  final BorderSide border;
```
I think rather than a BorderSide I would take a ShapeBorder (and default to `const StadiumBorder(side: const BorderSide.zero)`). Eventually we'll be moving most of our widgets to taking a ShapeBorder, for flexibility.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

gspencergoog committed:
More review changes
